Add pre post hooks

cwikla opened the pull request:
This adds

up(Pre|Post)Transaction
down(Pre|Post)Transaction

hooks, that when run in "each" mode will do the right thing, as in, run something not in a transaction. This is for very specific things like updating enums - it really shouldn't be used for anything else that I can think of?

Also made it so the CLI will use a new parameter from the connection config cli.transaction that matches the transaction mode of "all" | "each" | "none" that allows us to set the cli to use "each".

This is the typeorm half, I have tests and changing the GW migrations to use each in my GW branch.

cwikla committed:
Add pre/post hooks that get called outside of transactions

cwikla committed:
Add in connectionOptions.cli.transaction for setting the transaction mode

cwikla committed:
Fix up command line

mgromowsky commented on src/commands/MigrationRunCommand.ts:
```diff
@@ -58,7 +58,8 @@ export class MigrationRunCommand implements yargs.CommandModule {
             };
 
-            switch (args.t) {
+            const transactionMode = args.t !== "default" ? args.t : connectionOptions.cli?.transaction;
```
this line is confusing to me, but i'll take your word for it

cwikla replied:
if you look above in the initalizer, it gets set to the value "default" and then the switch statement drops it through. Why they used "default" instead of undefined? No idea. So basically if the command line has set the mode to anything besides default use it, otherwise try to use the one from the conneciton options. If that isn't set, it is undefined, and again, in the switch below, just falls through.

cwikla committed:
Move saving to db that migration has run until after the hook steps.

  The caveat is that because the post hook is run AFTER that we can
  now get into a state that the db is updated in the transaction, the
  migration fails in the post step and now we can't run the migration.

  Hmmm - probably need to track which steps have been run in the table.
